File: backend/app/services/bim_quantity_rules.py
from __future__ import annotations

import re
from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from typing import Any

QuantityRuleStatus = str


@dataclass(frozen=True)
class RuleDefinition:
    measure: str
    rule_hint: str
    units: tuple[str, ...]
    element_label: str = ""
    source: str = "system_default"
    allow_fallback_count: bool = False


CLASS_RULES: dict[str, RuleDefinition] = {
    "IFCBEAM": RuleDefinition("volumen o longitud", "NetVolume / NetLength", ("m3", "m")),
    "IFCBUILDINGELEMENTPROXY": RuleDefinition("conteo validado", "ElementCount", ("ea",), allow_fallback_count=True),
    "IFCCOLUMN": RuleDefinition("volumen o longitud", "NetVolume / NetLength", ("m3", "m")),
    "IFCCURTAINWALL": RuleDefinition("area", "NetSideArea / GrossSideArea", ("m2",)),
    "IFCDOOR": RuleDefinition("conteo", "Count / ElementCount", ("ea",), allow_fallback_count=True),
    "IFCFLOWFITTING": RuleDefinition("conteo", "Count / ElementCount", ("ea",), allow_fallback_count=True),
    "IFCFLOWSEGMENT": RuleDefinition("longitud", "NetLength", ("m",)),
    "IFCFLOWTERMINAL": RuleDefinition("conteo", "Count / ElementCount", ("ea",), allow_fallback_count=True),
    "IFCFOOTING": RuleDefinition("volumen", "NetVolume", ("m3",)),
    "IFCFURNISHINGELEMENT": RuleDefinition("conteo", "Count / ElementCount", ("ea",), allow_fallback_count=True),
    "IFCMEMBER": RuleDefinition("longitud", "NetLength", ("m",)),
    "IFCPILE": RuleDefinition("longitud o volumen", "NetLength / NetVolume", ("m", "m3")),
    "IFCPIPEFITTING": RuleDefinition("conteo", "Count / ElementCount", ("ea",), allow_fallback_count=True),
    "IFCPIPESEGMENT": RuleDefinition("longitud", "NetLength", ("m",)),
    "IFCPLATE": RuleDefinition("area", "NetArea / GrossArea", ("m2",)),
    "IFCRAILING": RuleDefinition("longitud", "NetLength", ("m",)),
    "IFCROOF": RuleDefinition("area o volumen", "NetArea / NetVolume", ("m2", "m3")),
    "IFCSLAB": RuleDefinition("area o volumen", "NetArea / NetVolume", ("m2", "m3")),
    "IFCSPACE": RuleDefinition("area o volumen", "NetFloorArea / NetVolume", ("m2", "m3")),
    "IFCSTAIR": RuleDefinition("volumen", "NetVolume", ("m3",)),
    "IFCWALL": RuleDefinition("area o volumen o longitud", "NetSideArea / NetVolume / NetLength", ("m2", "m3", "m")),
    "IFCWALLSTANDARDCASE": RuleDefinition(
        "area o volumen o longitud", "NetSideArea / NetVolume / NetLength", ("m2", "m3", "m")
    ),
    "IFCWINDOW": RuleDefinition("conteo", "Count / ElementCount", ("ea",), allow_fallback_count=True),
}


CLASS_LABELS: dict[str, str] = {
    "IFCBEAM": "Viga",
    "IFCBUILDINGELEMENTPROXY": "Elemento generico",
    "IFCCOLUMN": "Columna",
    "IFCCURTAINWALL": "Muro cortina",
    "IFCDOOR": "Puerta",
    "IFCFLOWFITTING": "Accesorio MEP",
    "IFCFLOWSEGMENT": "Tramo MEP",
    "IFCFLOWTERMINAL": "Terminal MEP",
    "IFCFOOTING": "Cimentacion",
    "IFCFURNISHINGELEMENT": "Mobiliario",
    "IFCMEMBER": "Miembro estructural",
    "IFCPILE": "Pilote",
    "IFCPIPEFITTING": "Accesorio de tuberia",
    "IFCPIPESEGMENT": "Tuberia",
    "IFCPLATE": "Placa",
    "IFCRAILING": "Baranda",
    "IFCROOF": "Cubierta",
    "IFCSLAB": "Losa",
    "IFCSPACE": "Espacio",
    "IFCSTAIR": "Escalera",
    "IFCWALL": "Muro",
    "IFCWALLSTANDARDCASE": "Muro",
    "IFCWINDOW": "Ventana",
}


def evaluate_quantity_rule(
    line: Mapping[str, Any] | object,
    catalog: Mapping[str, RuleDefinition] | None = None,
) -> dict[str, Any]:
    line = effective_quantity_line(line)
    ifc_class = _field(line, "ifc_class") or _field(line, "category")
    definition = _rule_for(ifc_class, _field(line, "unit"), catalog)
    source = _source_for(line)
    findings: list[str] = []
    normalized_unit = normalize_unit(_field(line, "unit"))
    expected_units = list(definition.units)
    preferred_measure = definition.measure.split(" o ")[0].strip() or definition.measure
    preferred_unit = expected_units[0] if expected_units else ""
    inferred_rule_unit = _measurement_unit(_field(line, "measurement_rule"))
    fallback_count_allowed = definition.allow_fallback_count and "ea" in expected_units
    quantity = _float_field(line, "quantity")

    if quantity <= 0:
        findings.append("Cantidad debe ser mayor que cero.")
    if not normalized_unit:
        findings.append("Unidad pendiente.")
    if expected_units and normalized_unit and normalized_unit not in expected_units:
        findings.append(
            f"Unidad {_field(line, 'unit')} no coincide con la regla esperada ({' / '.join(expected_units)})."
        )
    if inferred_rule_unit and normalized_unit and inferred_rule_unit != normalized_unit:
        findings.append(
            f"La regla {_field(line, 'measurement_rule')} requiere {inferred_rule_unit}, no {_field(line, 'unit')}."
        )
    if source == "Conteo fallback" and not fallback_count_allowed:
        findings.append(
            f"Medicion dimensional requerida: {_field(line, 'ifc_class') or 'la clase IFC'} "
            f"debe medirse por {preferred_measure} ({preferred_unit})."
        )

    status: QuantityRuleStatus = "blocked" if findings else "valid"
    confidence = (
        "Alta"
        if status == "valid" and source in {"IFC Quantity Set publicado", "Calculo geometrico desde IFC"}
        else "Media"
    )

    display_class = _field(line, "ifc_class") or "Clase IFC pendiente"
    measurement_rule = _field(line, "measurement_rule") or "Sin regla"
    source_text = "conteo de elementos sin Quantity Set" if source == "Conteo fallback" else source
    return {
        "allow_fallback_count": fallback_count_allowed,
        "status": status,
        "confidence": confidence,
        "source": source,
        "policy_version": 2,
        "rule_source": definition.source,
        "element_label": definition.element_label or CLASS_LABELS.get(normalize_ifc_class(ifc_class), display_class),
        "expected_measure": definition.measure,
        "expected_units": expected_units,
        "preferred_measure": preferred_measure,
        "preferred_unit": preferred_unit,
        "accepted_rules": [item.strip() for item in definition.rule_hint.split("/")],
        "findings": findings,
        "explanation": (
            f"{display_class}: regla esperada {definition.measure}; {definition.rule_hint}. "
            f"{measurement_rule} viene de {source_text}."
        ),
    }


def summarize_quantity_rules(
    lines: list[Mapping[str, Any] | object],
    catalog: Mapping[str, RuleDefinition] | None = None,
) -> dict[str, int]:
    summary = {"authoritative": 0, "blocked": 0, "review": 0, "total": 0, "valid": 0}
    for line in lines:
        result = evaluate_quantity_rule(line, catalog)
        summary["total"] += 1
        summary[result["status"]] += 1
        if result["source"] == "IFC Quantity Set publicado" and result["status"] == "valid":
            summary["authoritative"] += 1
    return summary


def normalize_ifc_class(value: str) -> str:
    return re.sub(r"[^A-Za-z0-9]+", "", _compact(value)).upper()


def normalize_unit(value: str | None) -> str:
    text = _compact(str(value or "")).lower().replace(" ", "")
    if text in {"m2", "sqm", "sq.m", "m^2"}:
        return "m2"
    if text in {"m3", "cum", "cu.m", "m^3"}:
        return "m3"
    if text in {"ea", "each", "und", "u", "un", "unidad", "unidades"}:
        return "ea"
    if text in {"ml", "lm", "m"}:
        return "m"
    if text in {"kg", "kilogram", "kilograms"}:
        return "kg"
    return text


def default_quantity_rule_records() -> list[dict[str, Any]]:
    return [
        {
            "ifc_class": ifc_class,
            "element_label": CLASS_LABELS.get(ifc_class, _humanize_ifc_class(ifc_class)),
            "expected_measure": rule.measure,
            "rule_hint": rule.rule_hint,
            "expected_units": list(rule.units),
            "allow_fallback_count": rule.allow_fallback_count,
            "source": "system_default",
            "status": "active",
        }
        for ifc_class, rule in sorted(CLASS_RULES.items())
    ]


def build_quantity_rule_catalog(records: Iterable[Mapping[str, Any] | object]) -> dict[str, RuleDefinition]:
    catalog: dict[str, RuleDefinition] = {}
    for record in records:
        status = _field(record, "status") or "active"
        if status != "active":
            continue
        ifc_class = normalize_ifc_class(_field(record, "ifc_class"))
        if not ifc_class:
            continue
        units = tuple(filter(None, (normalize_unit(unit) for unit in _list_field(record, "expected_units"))))
        catalog[ifc_class] = RuleDefinition(
            measure=_field(record, "expected_measure") or "cantidad controlada",
            rule_hint=_field(record, "rule_hint") or "Quantity / Unit",
            units=units,
            element_label=_field(record, "element_label")
            or CLASS_LABELS.get(ifc_class, _humanize_ifc_class(ifc_class)),
            source=_field(record, "source") or "project",
            allow_fallback_count=_bool_field(record, "allow_fallback_count", False),
        )
    return catalog


def _rule_for(
    ifc_class: str,
    unit: str,
    catalog: Mapping[str, RuleDefinition] | None = None,
) -> RuleDefinition:
    normalized = normalize_ifc_class(ifc_class)
    if catalog and normalized in catalog:
        return catalog[normalized]
    return CLASS_RULES.get(
        normalized,
        RuleDefinition("cantidad controlada", "Quantity / Unit", tuple(filter(None, [normalize_unit(unit)]))),
    )


def effective_quantity_line(line: Mapping[str, Any] | object) -> dict[str, Any]:
    raw_data = dict(_raw_data(line))
    controlled = raw_data.get("controlled_measurement")
    controlled_record = controlled if isinstance(controlled, dict) else {}
    use_controlled = controlled_record.get("status") == "approved"
    fields = (
        "ifc_class",
        "category",
        "quantity",
        "unit",
        "measurement_rule",
        "validation_notes",
    )
    payload: dict[str, Any] = {}
    for field in fields:
        if isinstance(line, Mapping):
            payload[field] = line.get(field, "")
        else:
            payload[field] = getattr(line, field, "")
    payload["raw_data"] = raw_data
    if use_controlled:
        payload["quantity"] = controlled_record.get("quantity", payload["quantity"])
        payload["unit"] = controlled_record.get("unit", payload["unit"])
        payload["measurement_rule"] = controlled_record.get("measurement_rule", payload["measurement_rule"])
    return payload


def evaluate_effective_quantity_rule(
    line: Mapping[str, Any] | object,
    catalog: Mapping[str, RuleDefinition] | None = None,
) -> dict[str, Any]:
    return evaluate_quantity_rule(effective_quantity_line(line), catalog)


def _measurement_unit(measurement_rule: str) -> str:
    normalized = _compact(measurement_rule).lower()
    if "volume" in normalized:
        return "m3"
    if "area" in normalized:
        return "m2"
    if "length" in normalized:
        return "m"
    if "count" in normalized:
        return "ea"
    return ""


def _source_for(line: Mapping[str, Any] | object) -> str:
    raw_data = _raw_data(line)
    controlled = raw_data.get("controlled_measurement")
    if isinstance(controlled, dict) and controlled.get("status") == "approved":
        controlled_rule = _compact(str(controlled.get("measurement_rule") or "")).lower()
        controlled_source = _compact(str(controlled.get("source") or "")).lower()
        if controlled_rule.startswith("geometry") or "geometry" in controlled_source or "geometr" in controlled_source:
            return "Calculo geometrico desde IFC"
        return "Plantilla Excel/CSV controlada"
    measurement_rule = _field(line, "measurement_rule").lower()
    notes = _field(line, "validation_notes").lower()
    if measurement_rule == "elementcount" or "no published ifc quantity" in notes:
        return "Conteo fallback"
    if _compact(str(raw_data.get("ifc_entity", ""))):
        return "IFC Quantity Set publicado"
    return "Plantilla Excel/CSV controlada"


def _field(line: Mapping[str, Any] | object, name: str) -> str:
    if isinstance(line, Mapping):
        value = line.get(name, "")
    else:
        value = getattr(line, name, "")
    if isinstance(value, list):
        return "; ".join(str(item) for item in value)
    return _compact(str(value or ""))


def _float_field(line: Mapping[str, Any] | object, name: str) -> float:
    try:
        return float(_field(line, name) or 0)
    except ValueError:
        return 0.0


def _raw_data(line: Mapping[str, Any] | object) -> dict[str, Any]:
    if isinstance(line, Mapping):
        value = line.get("raw_data", {})
    else:
        value = getattr(line, "raw_data", {})
    return value if isinstance(value, dict) else {}


def _list_field(line: Mapping[str, Any] | object, name: str) -> list[str]:
    if isinstance(line, Mapping):
        value = line.get(name, [])
    else:
        value = getattr(line, name, [])
    if isinstance(value, (list, tuple)):
        return [_compact(str(item)) for item in value if _compact(str(item))]
    text = _compact(str(value or ""))
    if not text:
        return []
    return [_compact(item) for item in re.split(r"[,/;]+", text) if _compact(item)]


def _bool_field(line: Mapping[str, Any] | object, name: str, default: bool = False) -> bool:
    if isinstance(line, Mapping):
        value = line.get(name, default)
    else:
        value = getattr(line, name, default)
    if isinstance(value, bool):
        return value
    return _compact(str(value)).lower() in {"1", "true", "yes", "si"}


def _humanize_ifc_class(value: str) -> str:
    text = re.sub(r"^IFC", "", value.upper()).title()
    return re.sub(r"([a-z])([A-Z])", r"\1 \2", text).strip() or value


def _compact(value: str) -> str:
    return value.strip()

File: frontend/src/lib/bimQuantityRules.ts
import type { QuantityTakeoffLine } from "../types";

export type QuantityRuleStatus = "valid" | "review" | "blocked";

export type QuantityRuleResult = {
  allowFallbackCount: boolean;
  confidence: "Alta" | "Media" | "Baja";
  expectedMeasure: string;
  expectedUnits: string[];
  explanation: string;
  findings: string[];
  policyVersion: number;
  preferredMeasure: string;
  preferredUnit: string;
  source:
    | "IFC Quantity Set publicado"
    | "Plantilla Excel/CSV controlada"
    | "Conteo fallback"
    | "Calculo geometrico desde IFC";
  status: QuantityRuleStatus;
};

export type QuantityRuleSummary = {
  authoritative: number;
  blocked: number;
  review: number;
  total: number;
  valid: number;
};

export type IfcGeometryDimensions = {
  x: number;
  y: number;
  z: number;
};

export type GeometricQuantityEstimate = {
  confidence: "Media";
  explanation: string;
  measurementRule: "GeometryAreaBBox" | "GeometryLengthBBox" | "GeometryVolumeBBox";
  quantity: number;
  source: "Calculo geometrico desde IFC";
  unit: "m" | "m2" | "m3";
};

type RuleDefinition = {
  allowFallbackCount?: boolean;
  measure: string;
  ruleHint: string;
  units: string[];
};

const CLASS_RULES: Record<string, RuleDefinition> = {
  IFCBEAM: { measure: "volumen o longitud", ruleHint: "NetVolume / NetLength", units: ["m3", "m"] },
  IFCBUILDINGELEMENTPROXY: {
    allowFallbackCount: true,
    measure: "conteo validado",
    ruleHint: "ElementCount",
    units: ["ea", "und"],
  },
  IFCCOLUMN: { measure: "volumen o longitud", ruleHint: "NetVolume / NetLength", units: ["m3", "m"] },
  IFCCURTAINWALL: { measure: "area", ruleHint: "NetSideArea / GrossSideArea", units: ["m2"] },
  IFCDOOR: { allowFallbackCount: true, measure: "conteo", ruleHint: "Count / ElementCount", units: ["ea", "und"] },
  IFCFLOWFITTING: {
    allowFallbackCount: true,
    measure: "conteo",
    ruleHint: "Count / ElementCount",
    units: ["ea", "und"],
  },
  IFCFLOWSEGMENT: { measure: "longitud", ruleHint: "NetLength", units: ["m"] },
  IFCFLOWTERMINAL: {
    allowFallbackCount: true,
    measure: "conteo",
    ruleHint: "Count / ElementCount",
    units: ["ea", "und"],
  },
  IFCFOOTING: { measure: "volumen", ruleHint: "NetVolume", units: ["m3"] },
  IFCMEMBER: { measure: "longitud", ruleHint: "NetLength", units: ["m"] },
  IFCPILE: { measure: "longitud o volumen", ruleHint: "NetLength / NetVolume", units: ["m", "m3"] },
  IFCPIPEFITTING: {
    allowFallbackCount: true,
    measure: "conteo",
    ruleHint: "Count / ElementCount",
    units: ["ea", "und"],
  },
  IFCPIPESEGMENT: { measure: "longitud", ruleHint: "NetLength", units: ["m"] },
  IFCPLATE: { measure: "area", ruleHint: "NetArea / GrossArea", units: ["m2"] },
  IFCRAILING: { measure: "longitud", ruleHint: "NetLength", units: ["m"] },
  IFCROOF: { measure: "area o volumen", ruleHint: "NetArea / NetVolume", units: ["m2", "m3"] },
  IFCSLAB: { measure: "area o volumen", ruleHint: "NetArea / NetVolume", units: ["m2", "m3"] },
  IFCSPACE: { measure: "area o volumen", ruleHint: "NetFloorArea / NetVolume", units: ["m2", "m3"] },
  IFCSTAIR: { measure: "volumen", ruleHint: "NetVolume", units: ["m3"] },
  IFCWALL: {
    measure: "area o volumen o longitud",
    ruleHint: "NetSideArea / NetVolume / NetLength",
    units: ["m2", "m3", "m"],
  },
  IFCWALLSTANDARDCASE: {
    measure: "area o volumen o longitud",
    ruleHint: "NetSideArea / NetVolume / NetLength",
    units: ["m2", "m3", "m"],
  },
  IFCWINDOW: { allowFallbackCount: true, measure: "conteo", ruleHint: "Count / ElementCount", units: ["ea", "und"] },
};

function compact(value: string | null | undefined) {
  return value?.trim() ?? "";
}

function normalizeIfcClass(value: string) {
  return compact(value)
    .replace(/[^A-Za-z0-9]/g, "")
    .toUpperCase();
}

function normalizeLengthScaleToMeters(value: string | null | undefined) {
  const text = compact(value)
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "");
  if (!text) return null;
  if (["mm", "milli", "millimeter", "millimeters", "millimetre", "millimetres"].includes(text)) return 0.001;
  if (["cm", "centi", "centimeter", "centimeters", "centimetre", "centimetres"].includes(text)) return 0.01;
  if (["m", "meter", "meters", "metre", "metres"].includes(text)) return 1;
  if (text.includes("millimeter") || text.includes("millimetre")) return 0.001;
  if (text.includes("centimeter") || text.includes("centimetre")) return 0.01;
  if (text.includes("meter") || text.includes("metre")) return 1;
  return null;
}

function normalizeUnit(value: string) {
  const text = compact(value).toLowerCase().replace(/\s+/g, "");
  if (["m2", "m²", "sqm", "sq.m", "m^2"].includes(text)) return "m2";
  if (["m3", "m³", "cum", "cu.m", "m^3"].includes(text)) return "m3";
  if (["ea", "each", "und", "u", "un", "unidad", "unidades"].includes(text)) return "ea";
  if (["ml", "lm", "m"].includes(text)) return "m";
  if (["kg", "kilogram", "kilograms"].includes(text)) return "kg";
  return text;
}

function roundQuantity(value: number) {
  return Number(value.toFixed(3));
}

function metricDimensions(dimensions: IfcGeometryDimensions | null | undefined, units: string | null | undefined) {
  if (!dimensions) return null;
  const scale = normalizeLengthScaleToMeters(units);
  if (!scale) return null;
  const values = [dimensions.x, dimensions.y, dimensions.z].map((value) => Number(value) * scale);
  if (values.some((value) => !Number.isFinite(value) || value <= 0)) return null;
  return values.sort((left, right) => right - left);
}

export function estimateQuantityFromGeometry(
  ifcClass: string,
  dimensions: IfcGeometryDimensions | null | undefined,
  modelUnits?: string | null
): GeometricQuantityEstimate | null {
  const normalizedClass = normalizeIfcClass(ifcClass);
  const dims = metricDimensions(dimensions, modelUnits);
  if (!normalizedClass || !dims) return null;

  const areaClasses = new Set([
    "IFCCURTAINWALL",
    "IFCPLATE",
    "IFCROOF",
    "IFCSLAB",
    "IFCSPACE",
    "IFCWALL",
    "IFCWALLSTANDARDCASE",
  ]);
  const volumeClasses = new Set(["IFCBEAM", "IFCCOLUMN", "IFCFOOTING", "IFCPILE", "IFCSTAIR"]);
  const lengthClasses = new Set(["IFCFLOWSEGMENT", "IFCMEMBER", "IFCPIPESEGMENT", "IFCRAILING"]);

  if (areaClasses.has(normalizedClass)) {
    return {
      confidence: "Media",
      explanation: "Area sugerida con las dos mayores dimensiones del bounding box IFC seleccionado.",
      measurementRule: "GeometryAreaBBox",
      quantity: roundQuantity(dims[0] * dims[1]),
      source: "Calculo geometrico desde IFC",
      unit: "m2",
    };
  }
  if (volumeClasses.has(normalizedClass)) {
    return {
      confidence: "Media",
      explanation: "Volumen sugerido con largo, ancho y alto del bounding box IFC seleccionado.",
      measurementRule: "GeometryVolumeBBox",
      quantity: roundQuantity(dims[0] * dims[1] * dims[2]),
      source: "Calculo geometrico desde IFC",
      unit: "m3",
    };
  }
  if (lengthClasses.has(normalizedClass)) {
    return {
      confidence: "Media",
      explanation: "Longitud sugerida con la mayor dimension del bounding box IFC seleccionado.",
      measurementRule: "GeometryLengthBBox",
      quantity: roundQuantity(dims[0]),
      source: "Calculo geometrico desde IFC",
      unit: "m",
    };
  }

  return null;
}

function controlledMeasurement(line: QuantityTakeoffLine) {
  const value = line.raw_data?.controlled_measurement;
  return value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : null;
}

function effectiveQuantityValues(line: QuantityTakeoffLine) {
  const controlled = controlledMeasurement(line);
  if (controlled && String(controlled.status ?? "") === "approved") {
    return {
      measurementRule: compact(String(controlled.measurement_rule ?? line.measurement_rule)),
      quantity: Number(controlled.quantity ?? line.quantity),
      unit: compact(String(controlled.unit ?? line.unit)),
    };
  }
  return { measurementRule: line.measurement_rule, quantity: Number(line.quantity || 0), unit: line.unit };
}

function sourceFor(line: QuantityTakeoffLine) {
  const controlled = controlledMeasurement(line);
  if (controlled && String(controlled.status ?? "") === "approved") {
    const rule = compact(String(controlled.measurement_rule ?? "")).toLowerCase();
    const source = compact(String(controlled.source ?? "")).toLowerCase();
    if (rule.startsWith("geometry") || source.includes("geometry") || source.includes("geometr")) {
      return "Calculo geometrico desde IFC" as const;
    }
    return "Plantilla Excel/CSV controlada" as const;
  }
  const rule = compact(line.measurement_rule).toLowerCase();
  const notes = compact(line.validation_notes).toLowerCase();
  if (rule === "elementcount" || notes.includes("no published ifc quantity")) {
    return "Conteo fallback" as const;
  }
  if (line.raw_data && typeof line.raw_data === "object" && compact(String(line.raw_data.ifc_entity ?? ""))) {
    return "IFC Quantity Set publicado" as const;
  }
  return "Plantilla Excel/CSV controlada" as const;
}

function ruleFor(line: QuantityTakeoffLine) {
  const ifcClass = normalizeIfcClass(line.ifc_class || line.category);
  return (
    CLASS_RULES[ifcClass] ?? {
      measure: "cantidad controlada",
      ruleHint: "Quantity / Unit",
      units: [normalizeUnit(line.unit)].filter(Boolean),
    }
  );
}

function preferredMeasure(definition: RuleDefinition) {
  return definition.measure.split(" o ")[0]?.trim() || definition.measure;
}

function measurementUnit(measurementRule: string) {
  const normalized = compact(measurementRule).toLowerCase();
  if (normalized.includes("volume")) return "m3";
  if (normalized.includes("area")) return "m2";
  if (normalized.includes("length")) return "m";
  if (normalized.includes("count")) return "ea";
  return "";
}

function auditedRuleFor(line: QuantityTakeoffLine): QuantityRuleResult | null {
  const rawRule = line.raw_data?.quantity_rule;
  if (!rawRule || typeof rawRule !== "object" || Array.isArray(rawRule)) return null;
  const rule = rawRule as Record<string, unknown>;
  if (Number(rule.policy_version ?? 0) < 2) return null;
  const status = String(rule.status ?? "");
  const confidence = String(rule.confidence ?? "");
  const source = String(rule.source ?? "");
  if (!["valid", "review", "blocked"].includes(status)) return null;
  if (!["Alta", "Media", "Baja"].includes(confidence)) return null;
  if (
    ![
      "IFC Quantity Set publicado",
      "Plantilla Excel/CSV controlada",
      "Conteo fallback",
      "Calculo geometrico desde IFC",
    ].includes(source)
  )
    return null;

  return {
    allowFallbackCount: Boolean(rule.allow_fallback_count),
    confidence: confidence as QuantityRuleResult["confidence"],
    expectedMeasure: String(rule.expected_measure ?? rule.expectedMeasure ?? ""),
    expectedUnits: Array.isArray(rule.expected_units)
      ? rule.expected_units.map((unit) => String(unit)).filter(Boolean)
      : Array.isArray(rule.expectedUnits)
        ? rule.expectedUnits.map((unit) => String(unit)).filter(Boolean)
        : [],
    explanation: String(rule.explanation ?? ""),
    findings: Array.isArray(rule.findings) ? rule.findings.map((finding) => String(finding)).filter(Boolean) : [],
    policyVersion: Number(rule.policy_version ?? 2),
    preferredMeasure: String(rule.preferred_measure ?? ""),
    preferredUnit: String(rule.preferred_unit ?? ""),
    source: source as QuantityRuleResult["source"],
    status: status as QuantityRuleStatus,
  };
}

export function evaluateQuantityRule(line: QuantityTakeoffLine): QuantityRuleResult {
  const auditedRule = auditedRuleFor(line);
  if (auditedRule) return auditedRule;

  const definition = ruleFor(line);
  const source = sourceFor(line);
  const effective = effectiveQuantityValues(line);
  const findings: string[] = [];
  const normalizedUnit = normalizeUnit(effective.unit);
  const expectedUnits = definition.units;
  const allowFallbackCount = Boolean(
    definition.allowFallbackCount && expectedUnits.some((unit) => normalizeUnit(unit) === "ea")
  );
  const preferredUnit = normalizeUnit(expectedUnits[0] ?? "");
  const inferredRuleUnit = measurementUnit(effective.measurementRule);

  if (!effective.quantity || effective.quantity <= 0) findings.push("Cantidad debe ser mayor que cero.");
  if (!normalizedUnit) findings.push("Unidad pendiente.");
  if (expectedUnits.length && normalizedUnit && !expectedUnits.map(normalizeUnit).includes(normalizedUnit)) {
    findings.push(`Unidad ${effective.unit} no coincide con la regla esperada (${expectedUnits.join(" / ")}).`);
  }
  if (inferredRuleUnit && normalizedUnit && inferredRuleUnit !== normalizedUnit) {
    findings.push(`La regla ${effective.measurementRule} requiere ${inferredRuleUnit}, no ${effective.unit}.`);
  }
  if (source === "Conteo fallback" && !allowFallbackCount) {
    findings.push(
      `Medicion dimensional requerida: ${compact(line.ifc_class) || "la clase IFC"} debe medirse por ${preferredMeasure(definition)} (${preferredUnit}).`
    );
  }

  const hasBlockingFinding = findings.length > 0;
  const status: QuantityRuleStatus = hasBlockingFinding ? "blocked" : "valid";
  const confidence =
    status === "blocked"
      ? "Media"
      : source === "IFC Quantity Set publicado"
        ? "Alta"
        : source === "Calculo geometrico desde IFC"
          ? "Alta"
          : source === "Plantilla Excel/CSV controlada"
            ? "Media"
            : "Media";
  const ifcClass = compact(line.ifc_class) || "Clase IFC pendiente";
  const sourceText = source === "Conteo fallback" ? "conteo de elementos sin Quantity Set" : source;

  return {
    allowFallbackCount,
    confidence,
    expectedMeasure: definition.measure,
    expectedUnits,
    explanation: `${ifcClass}: regla esperada ${definition.measure}; ${definition.ruleHint}. ${effective.measurementRule || "Sin regla"} viene de ${sourceText}.`,
    findings,
    policyVersion: 2,
    preferredMeasure: preferredMeasure(definition),
    preferredUnit,
    source,
    status,
  };
}

export function summarizeQuantityRules(lines: QuantityTakeoffLine[]): QuantityRuleSummary {
  return lines.reduce<QuantityRuleSummary>(
    (summary, line) => {
      const result = evaluateQuantityRule(line);
      summary.total += 1;
      if (result.status === "valid") summary.valid += 1;
      if (result.status === "review") summary.review += 1;
      if (result.status === "blocked") summary.blocked += 1;
      if (result.source === "IFC Quantity Set publicado" && result.status === "valid") summary.authoritative += 1;
      return summary;
    },
    { authoritative: 0, blocked: 0, review: 0, total: 0, valid: 0 }
  );
}
